runtime/pallene_core.h

```diff
@@ -226,4 +226,41 @@ int pallene_is_truthy(const TValue *v)
     return !(ttisnil(v) || (ttisboolean(v) && bvalue(v) == 0));
 }
 
+static inline
+Table *pallene_new_table(lua_State *L, lua_Integer n)
+{
+    Table *t = luaH_new(L);
+    if (n > 0) {
+        luaH_resize(L, t, 0, n);
+    }
+    return t;
+}
+
+static const TValue PALLENE_ABSENTKEY = {ABSTKEYCONSTANT};
+
+static inline
+TValue *pallene_getstr(Table *t, TString *key, int *pos)
+{
+    if (PALLENE_LIKELY(*pos < sizenode(t))) {
+       Node *n = gnode(t, *pos);
+       if (PALLENE_LIKELY(keyisshrstr(n) && eqshrstr(keystrval(n), key)))
+           return gval(n);
+    }
+    int currpos = lmod(key->hash, sizenode(t));
+    Node *n = gnode(t, currpos);
+    for (;;) {
+        if (keyisshrstr(n) && eqshrstr(keystrval(n), key)) {
+            *pos = currpos;
+            return gval(n);
+        }
+        else {
+            int nx = gnext(n);
+            if (nx == 0)
+                return (TValue *)&PALLENE_ABSENTKEY;  /* not found */
```

Comment:
Would it make sense to update the inline cache in this not-found case? For example `*pos = -1`?

Reply:
I don't see the point.

Reply:
Consider a program where the key is not found once. Odds are, that it will continue being not found once we go through this line again with other tables. If the cache remains a valid index, every time we're going to start the table access with an optimistic access to the cache, which is going to fail. It would be better to just skip and go straight to the for loop.